pkg/controller/certificaterequests/sync.go

```diff
@@ -0,0 +1,239 @@
+	}
+	if err != nil {
+		return err
+	}
+	dbg.Info("Fetched issuer resource referenced by certificate request", "issuer_name", crCopy.Spec.IssuerRef.Name)
+
+	issuerReady := apiutil.IssuerHasCondition(issuerObj, v1alpha1.IssuerCondition{
+		Type:   v1alpha1.IssuerConditionReady,
+		Status: v1alpha1.ConditionTrue,
+	})
+	if !issuerReady {
+		c.recorder.Eventf(crCopy, corev1.EventTypeWarning, errorIssuerNotReady, "Issuer %s not ready", issuerObj.GetObjectMeta().Name)
+		return nil
+	}
+
+	i, err := c.issuerFactory.IssuerFor(issuerObj)
+	if err != nil {
+		c.recorder.Eventf(crCopy, corev1.EventTypeWarning, errorIssuerInit, "Internal error initialising issuer: %v", err)
+		return nil
+	}
+
+	el = validation.ValidateCertificateRequestForIssuer(crCopy, issuerObj)
+	if len(el) > 0 {
+		c.recorder.Eventf(crCopy, corev1.EventTypeWarning, "BadConfig", "Resource validation failed: %v", el.ToAggregate())
+		return nil
+	}
+
+	if len(cr.Status.Certificate) == 0 {
+		dbg.Info("Invoking sign function as existing certificate does not exist")
+		return c.sign(ctx, i, crCopy, csr)
+	}
+
+	cert, err := pki.DecodeX509CertificateBytes(cr.Status.Certificate)
+	if err != nil {
+		return err
+	}
+
+	dbg.Info("Update certificate request status if required")
+	c.setCertificateRequestStatus(crCopy, csr, cert)
+
+	// TODO: Metrics??
+
+	if cert == nil {
+		dbg.Info("Invoking sign function as existing certificate does not exist")
+		return c.sign(ctx, i, crCopy, csr)
+	}
+
+	dbg.Info("Certificate does not need updating.")
+
+	return nil
+}
+
+// return an error on failure. If retrieval is succesful, the certificate data
+// will be stored in the certificate request status
+func (c *controller) sign(ctx context.Context, issuer issuer.Interface,
+	cr *v1alpha1.CertificateRequest, csr *x509.CertificateRequest) error {
+	log := logf.FromContext(ctx)
+
+	resp, err := issuer.Sign(ctx, cr)
+	if err != nil {
+		log.Error(err, "error issuing certificate request")
+		return err
+	}
+
+	// if the issuer has not returned any data, exit early
+	if resp == nil {
+		c.setCertificateRequestStatus(cr, nil, nil)
+		return nil
+	}
+
+	if len(resp.Certificate) > 0 {
+		cr.Status.Certificate = resp.Certificate
+		cr.Status.CA = resp.CA
+
+		c.recorder.Event(cr, corev1.EventTypeNormal, successCertificateIssued, "Certificate issued successfully")
+	}
+
+	cert, err := pki.DecodeX509CertificateBytes(resp.Certificate)
+	if err != nil {
+		return err
+	}
+
+	c.setCertificateRequestStatus(cr, csr, cert)
+
+	return nil
+}
+
+// setCertificateRequestStatus will update the status subresource of the
+// certificate reques. It will not actually submit the resource to the
+// apiserver.
+func (c *controller) setCertificateRequestStatus(cr *v1alpha1.CertificateRequest, csr *x509.CertificateRequest, cert *x509.Certificate) {
+	if cert == nil {
+		apiutil.SetCertificateRequestCondition(cr, v1alpha1.CertificateRequestConditionReady, v1alpha1.ConditionFalse, "NotExists", "Certificate does not exist")
```

Comment:
I think we need to consider how temporary certificates work. Will they continue to be issued by the Certificates controller?